src/Features/Core/Portable/ConvertAnonymousTypeToClass/AbstractConvertAnonymousTypeToClassCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,412 @@
+            if (!token.Span.IntersectsWith(position))
+            {
+                return default;
+            }
+
+            if (!span.IsEmpty && span != token.Span)
+            {
+                // if there is a selection, it has to be of the whole token.
+                return default;
+            }
+
+            var anonymousObject = token.Parent as TAnonymousObjectCreationExpressionSyntax;
+            if (anonymousObject == null)
+            {
+                return default;
+            }
+
+            // The position/selection must be of the 'new' token of the anonymous object.
+            if (anonymousObject.GetFirstToken() != token)
+            {
+                return default;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var anonymousType = semanticModel.GetTypeInfo(anonymousObject, cancellationToken).Type as INamedTypeSymbol;
+
+            return (anonymousObject, anonymousType);
+        }
+
+        private async Task<Document> ConvertToClassAsync(
+            Document document, TextSpan span, CancellationToken cancellationToken)
+        {
+            var (anonymousObject, anonymousType) = await TryGetAnonymousObjectAsync(
+                document, span, cancellationToken).ConfigureAwait(false);
+
+            Debug.Assert(anonymousObject != null);
+            Debug.Assert(anonymousType != null);
+
+            var position = span.Start;
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var compilation = semanticModel.Compilation;
+
+            // Generate a unique name for the class we're creating.  We'll also add a rename
+            // annotation so the user can pick the right name for the type afterwards.
+            var className = NameGenerator.GenerateUniqueName(
+                "NewClass", n => semanticModel.LookupSymbols(position, name: n).IsEmpty);
+
+            // First, create the set of properties this class will have based on the properties
+            // the anonymous type has itself.
+            var (properties, propertyMap) = GenerateProperties(anonymousType);
+
+            // Next, generate the full class that will be used to replace all instances of this
+            // anonymous type.
+            var namedTypeSymbol = await GenerateFinalNamedTypeAsync(
+                document, className, properties, cancellationToken).ConfigureAwait(false);
+
+            var generator = SyntaxGenerator.GetGenerator(document);
+            var editor = new SyntaxEditor(root, generator);
+
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var containingMember = anonymousObject.FirstAncestorOrSelf<SyntaxNode>(syntaxFacts.IsMethodLevelMember) ?? anonymousObject;
```

Comment:
>?? anonymousObject; [](start = 116, length = 19)

What's the scenario where we fallback to the `?? anonymousObject` case? Should we bail out instead in such a case?

Reply:
We could potentially bail out.  I was worried about things like potentially having code at the top level.  I didn't want to assume that we would *always* be in a member.  But i still thought it would be fine to gneerate the class.   

Thoughts on what you would prefer here?  Should i bail and not even offer the feature?  or just be resilient and still allow it?

Reply:
What's a scenario that reaches that case?

---
In reply to: [198309468](https://github.com/dotnet/roslyn/pull/28066#discussion_r198309468) [](ancestors = 198309468)

Reply:
One (illegal) example would be be an anonymous type in a global attribute.  It will parse, and i don't want us to crash there.

--

However, in general, i'd like the feature to be available in any place where you could write an anonymous type (including in the future).  so that's why i went for resiliency here.